Content.Shared/Charges/Systems/SharedChargesSystem.cs

```diff
@@ -1,103 +1,216 @@
-    /// </summary>
-    public bool TryUseCharge(Entity<LimitedChargesComponent?> ent)
+    public void AddCharges(Entity<LimitedChargesComponent?> action, int addCharges)
     {
-        if (!Query.Resolve(ent, ref ent.Comp, false))
-            return true;
+        if (addCharges == 0)
+            return;
+
+        action.Comp ??= EnsureComp<LimitedChargesComponent>(action.Owner);
 
-        if (IsEmpty(ent, ent.Comp))
+        var oldCharges = GetCurrentCharges((action.Owner, action.Comp, null));
+        var charges = Math.Clamp(oldCharges + addCharges, 0, action.Comp.MaxCharges);
+
+        if (oldCharges == charges)
+            return;
+
+        action.Comp.LastCharges = charges;
+        action.Comp.LastUpdate = _timing.CurTime;
+        Dirty(action);
+    }
+
+    public bool TryUseCharge(Entity<LimitedChargesComponent?> entity)
+    {
+        return TryUseCharges(entity, 1);
+    }
+
+    public bool TryUseCharges(Entity<LimitedChargesComponent?> entity, int amount)
+    {
+        var current = GetCurrentCharges(entity);
+
+        if (current < amount)
+        {
             return false;
+        }
 
-        UseCharge(ent, ent.Comp);
+        AddCharges(entity, -amount);
         return true;
     }
 
+    [Pure]
+    public bool IsEmpty(Entity<LimitedChargesComponent?> entity)
+    {
+        return GetCurrentCharges(entity) == 0;
+    }
+
     /// <summary>
-    /// Gets the limited charges component and returns true if the number of charges remaining is less than the specified value.
-    /// Will return false if there is no limited charges component.
+    /// Resets action charges to MaxCharges.
     /// </summary>
-    public bool HasInsufficientCharges(EntityUid uid, int requiredCharges, LimitedChargesComponent? comp = null)
+    public void ResetCharges(Entity<LimitedChargesComponent?> action)
     {
-        // can't be empty if there are no limited charges
-        if (!Resolve(uid, ref comp, false))
-            return false;
+        if (!Resolve(action.Owner, ref action.Comp, false))
+            return;
 
-        return comp.Charges < requiredCharges;
+        var charges = GetCurrentCharges((action.Owner, action.Comp, null));
+
+        if (charges == action.Comp.MaxCharges)
+            return;
+
+        action.Comp.LastCharges = action.Comp.MaxCharges;
+        action.Comp.LastUpdate = _timing.CurTime;
+        Dirty(action);
+    }
+
+    public void SetCharges(Entity<LimitedChargesComponent?> action, int value)
+    {
+        action.Comp ??= EnsureComp<LimitedChargesComponent>(action.Owner);
+
+        var adjusted = Math.Clamp(value, 0, action.Comp.MaxCharges);
+
+        if (action.Comp.LastCharges == adjusted)
+        {
+            return;
+        }
+
+        action.Comp.LastCharges = adjusted;
+        action.Comp.LastUpdate = _timing.CurTime;
+        Dirty(action);
+    }
+
+    [Pure]
+    private TimeSpan GetNextRechargeTime(Entity<LimitedChargesComponent?, AutoRechargeComponent?> entity)
```

Comment:
```suggestion
    public TimeSpan GetNextRechargeTime(Entity<LimitedChargesComponent?, AutoRechargeComponent?> entity)
```

public for where this will be used.